File: test-suite/go/tests/valkey.go
package tests

import (
	"context"
	"fmt"

	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

const (
	valkeyService = "Valkey"
	valkeyClient  = "go-redis"
)

// TestValkey tests Valkey connectivity
func TestValkey(utils TestUtils) ServiceResult {
	host := utils.GetEnv("VALKEY_HOST", utils.GetEnv("REDIS_HOST", "valkey"))
	port := utils.GetEnvInt("VALKEY_PORT", utils.GetEnvInt("REDIS_PORT", 6379))

	if gate := utils.VerifyServiceGate(valkeyService, valkeyClient, "ENABLE_VALKEY", false, host, port); gate != nil {
		return *gate
	}

	return utils.RunCheck(valkeyService, valkeyClient, func() (string, error) {
		ctx := context.Background()
		client := redis.NewClient(&redis.Options{
			Addr: fmt.Sprintf("%s:%d", host, port),
		})
		defer client.Close()

		// Test connection
		payload := uuid.New().String()
		key := fmt.Sprintf("health:%s", payload)

		// Set value with expiration
		err := client.Set(ctx, key, payload, 0).Err()
		if err != nil {
			return "", err
		}

		// Get value
		value, err := client.Get(ctx, key).Result()
		if err != nil {
			return "", err
		}

		// Delete key
		client.Del(ctx, key)

		if value != payload {
			return "", fmt.Errorf("unexpected payload: got %s, want %s", value, payload)
		}

		return fmt.Sprintf("SET/GET on %s succeeded", key), nil
	})
}

File: test-suite/rust/src/tests/valkey.rs
use crate::utils::*;
use redis::Commands;
use uuid::Uuid;

const SERVICE: &str = "Valkey";
const CLIENT_NAME: &str = "rust-redis";

pub fn test_valkey() -> ServiceResult {
    let host = get_env("VALKEY_HOST", &get_env("REDIS_HOST", "valkey"));
    let port = get_env_u16("VALKEY_PORT", get_env_u16("REDIS_PORT", 6379));

    if let Some(gate) = verify_service_gate(SERVICE, CLIENT_NAME, "ENABLE_VALKEY", false, &host, port) {
        return gate;
    }

    run_check(SERVICE, CLIENT_NAME, || {
        let redis_url = format!("redis://{}:{}", host, port);
        let client = redis::Client::open(redis_url)?;
        let mut con = client.get_connection()?;

        // Test connection
        let payload = Uuid::new_v4().to_string();
        let key = format!("health:{}", payload);

        // Set value
        con.set::<_, _, ()>(&key, &payload)?;

        // Get value
        let value: String = con.get(&key)?;

        // Delete key
        con.del::<_, ()>(&key)?;

        if value != payload {
            return Err(format!("unexpected payload: got {}, want {}", value, payload).into());
        }

        Ok(format!("SET/GET on {} succeeded", key))
    })
}
